lib/events/filesessions/fileasync.go

```diff
@@ -128,49 +129,125 @@ type Uploader struct {
 	log             *log.Entry
 	uploadCompleter *events.UploadCompleter
 
-	cancel context.CancelFunc
-	ctx    context.Context
+	cancel   context.CancelFunc
+	ctx      context.Context
+	eventsCh chan events.UploadEvent
+}
+
+func (u *Uploader) writeSessionError(sessionID session.ID, err error) error {
+	if sessionID == "" {
+		return trace.BadParameter("missing session ID")
+	}
+	path := u.sessionErrorFilePath(sessionID)
+	return trace.ConvertSystemError(ioutil.WriteFile(path, []byte(err.Error()), 0600))
+}
+
+func (u *Uploader) checkSessionError(sessionID session.ID) (bool, error) {
+	if sessionID == "" {
+		return false, trace.BadParameter("missing session ID")
+	}
+	_, err := os.Stat(u.sessionErrorFilePath(sessionID))
+	if err != nil {
+		err = trace.ConvertSystemError(err)
+		if trace.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
 }
 
 // Serve runs the uploader until stopped
 func (u *Uploader) Serve() error {
-	t := u.cfg.Clock.NewTicker(u.cfg.ScanPeriod)
-	defer t.Stop()
+	backoff, err := utils.NewLinear(utils.LinearConfig{
+		Step:  u.cfg.ScanPeriod,
+		Max:   u.cfg.ScanPeriod * 100,
+		Clock: u.cfg.Clock,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// Start after scan period delay
+	backoff.Inc()
 	for {
 		select {
 		case <-u.ctx.Done():
 			return nil
-		case <-t.Chan():
+		case event := <-u.eventsCh:
+			switch {
+			case event.Error == nil:
+				backoff.ResetToDelay()
+			case isSessionError(event.Error):
+				u.log.WithError(event.Error).Warningf(
+					"Failed to read session recording %v, will skip future uploads.", event.SessionID)
```

Comment:
optional: don't wrap these logging calls https://github.com/golang/go/wiki/CodeReviewComments#line-length

Reply:
It's a bit easier for me to read tbh.

Reply:
I generally agree though :)